crates/bevy_ecs/src/change_detection.rs

```diff
@@ -0,0 +1,181 @@
+use crate::component::{Component, ComponentTicks};
+use bevy_reflect::Reflect;
+use std::ops::{Deref, DerefMut};
+
+pub trait ChangeDetectable {
```

Comment:
I think we will eventually want to split this up into two traits (names tbd):

```rust
trait ChangeDetectable {
    fn is_added(&self) -> bool;
    fn is_changed(&self) -> bool;
    fn set_changed(&mut self);
}

trait Underlying {
    type Target: ?Sized;
    fn underlying(&self) -> &Self::Target;
    fn underlying_mut(&mut self) -> &mut Self::Target;
    fn underlying_mut_untracked(&mut self) -> &mut Self::Target;
}
```

That way we can check "changed state" across all "observables" without needing to know what type is stored. This could feed in to the "reactive ecs" scenario.

Reply:
It would also allow us to remove Underlying from the prelude, which I think is a good call given how uncommon the use case is.

Reply:
Yea, I like this a lot.
However, should one of the traits depend on the other?
e.g.
```rust
trait Underlying: ChangeDetectable { .. }
```

Reply:
Optionally we can remove the `Underlying` trait in it's entirety. 
The only thing I really liked from this was the `underlying_mut_unchecked` I feel like can have some usecases.

Reply:
Hmm yeah lets just remove Underlying for now. I don't see any strong motivating use cases other than circumventing change detection, which I'm yet convinced is something we need to support.